Add Quality by Design User Story and Supporting Docs (#7)
camaraproject/QualityByDesign: documentation/API_documentation/QualityByDesign_User_Story.md

> # Quality-By-Design (QbD) User Story

Reviewer: ```suggestion
# Session Insights User Story
```

| **Item**                        | **Description**                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     |
| ------------------------------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |

> | **_Summary_**                   | As an application developer, I want to be able to share network KPI requirements with a Communication Service Provider (CSP) and stream these KPIs for real-time monitoring. This allows the CSP to track network performance and ensure it aligns with the specified requirements. If any deviations occur, the CSP can proactively suggest or implement corrective measures to maintain optimal performance.                                                                                                                      |

Reviewer: I think the first sentence could be rephrased to highlight the difference with Connectivity Insights (which also involves sharing of KPIs): so suggest to focus on the stream aspect, e.g. 

"As an application developer, I want to be able to stream my application's network KPI requirements to the network operator. This allows the network operator to monitor in real-time whether it can fulfil these dynamic requirements, and if not, proactively suggest or implement corrective measures to meet the application's requirements."

(also in the above I have suggested to change 'CSP' to 'network operator' as it's common in CAMARA, and 'maintain optimal performance' to 'meet application requirements', since we don't know whether performance was optimal in the first place, and I don't believe we are targeting optimal so much as that which the application developer and end user are happy with, or better.

Author: thanks Kevin, good feedback! I've incorporated and will update the PR soon.

> | **_Roles, Actor(s) and scope_** | **Roles:** API Consumer:Developer<br><br>**Actors:** Application service providers, hyperscalers, application developers.<br><br>**Scope:**<ul><li>Create QbD Session</li><li>Send Network KPIs</li><li>Get QbD score</li><li>Get RCA and recommended corrective actions</li><li>End Session</li></ul>                                                                                                                                                                                                                              |

Reviewer: Should "Get RCA and recommended corrective actions" rather be ""Get Recommended Corrective Actions (RCA)"?

> | **_Roles, Actor(s) and scope_** | **Roles:** API Consumer:Developer<br><br>**Actors:** Application service providers, hyperscalers, application developers.<br><br>**Scope:**<ul><li>Create QbD Session</li><li>Send Network KPIs</li><li>Get QbD score</li><li>Get RCA and recommended corrective actions</li><li>End Session</li></ul>                                                                                                                                                                                                                              |

Author: RCA for me was Root Cause Analysis. I'll change it to "Get root cause and recommended corrective actions" to be explicit without acronyms.

> | **_Pre-conditions_**            | The preconditions are:<ol><li>Customer:Developer is eligible to use the QbD API, i.e. the Customer has successfully subscribed to the QbD API on the platform of an API Provider.</li><li>The customer application has requested and received an access token for the required scopes of the API.</li><li>The customer application determines which application profile it will be using for the QbD session.</li></ol>                                                                                                             |

Reviewer: >The customer application determines which application profile it will be using for the QbD session.

Or is the prerequisite rather that the application developer provisions an application profile first for that application?

Author: good point, will update

> | **_Begins when_**               | The customer application creates a QbD session using an application profile, device identifier or 3-legged token, and callback URI. The CSP provides a session ID in the response.                                                                                                                                                                                                                                                                                                                                                  |

Reviewer: ```suggestion
| **_Begins when_**               | The customer application creates a Session Insights session using an application profile, device identifier or 3-legged token, and callback URI. The CSP provides a session ID in the response.                                                                                                                                                                                                                                                                                                                                                  |
```

| **_Step 1_**                    | The customer application periodically sends its network KPIs for its session to the CSP.                                                                                                                                                                                                                                                                                                                                                                                                                                            |

> | **_Step 2_**                    | The CSP calculates a QbD score using KPIs and the application profile and sends it to the customer application via the callback URI provided.                                                                                                                                                                                                                                                                                                                                                                                       |

Reviewer: ```suggestion
| **_Step 2_**                    | The CSP calculates a Session Insights score using KPIs and the application profile and sends it to the customer application via the callback URI provided.                                                                                                                                                                                                                                                                                                                                                                                       |
```

| **_Step 3_**                    | The CSP sends asynchronous updates with the current score. The CSP can optionally include additional information to help diagnose and correct an issue when the score falls below acceptable thresholds.                                                                                                                                                                                                                                                                                                                            |

> | **_Step 4_**                    | When possible, the customer application takes corrective action.                                                                                                                                                                                                                                                                                                                                                                                                                                                                    |

Reviewer: I think this should rather be "The customer application decides whether to take corrective action"

Author: sure, I can change it to that. The point I was trying to make is that in some scenarios the issue may not be in the application, but lies in the network operator's domain. In that case there could be a resolution that takes place on the network and the application does not need to do anything.

> | **_Ends when_**                 | The customer application ends the QbD session or session times out due to inactivity.                                                                                                                                                                                                                                                                                                                                                                                                                                               |

Reviewer: ```suggestion
| **_Ends when_**                 | The customer application ends the Session Insights session or session times out due to inactivity.                                                                                                                                                                                                                                                                                                                                                                                                                                               |
```

> | **_Post-conditions_**           | The customer application was able to adjust to network issues in real-time and provide continued performance to the end user of the application.                                                                                                                                                                                                                                                                                                                                                                                    |

Reviewer: I think 'continued' should rather be 'consistent'

> | **_Exceptions_**                | Several exceptions might occur during the QbD API operations:<ul><li>Unauthorized: Not valid credentials (i.e. use of an expired access token).</li><li>Conflict: Internal configuration policies didn’t allow for operations, i.e. a second QbD session for one device.</li><li>Not Found: application profile provided for session is not a defined application profile</li><li>Bad Request: If KPIs provided are not valid</li><li>Session-not-created: Due to other internal errors, QbD session resource could not be created. |

Reviewer: ```suggestion
| **_Exceptions_**                | Several exceptions might occur during the SessionInsights API operations:<ul><li>Unauthorized: Not valid credentials (i.e. use of an expired access token).</li><li>Conflict: Internal configuration policies didn’t allow for operations, i.e. a second  SessionInsights session for one device.</li><li>Not Found: application profile provided for session is not a defined application profile</li><li>Bad Request: If KPIs provided are not valid</li><li>Session-not-created: Due to other internal errors,  SessionInsights session resource could not be created. |
```
